while-loops/while-loops.md

```diff
@@ -0,0 +1,257 @@
+        if i == 7:
+            print("i is seven! Let's get out of here!")
+            break
+        print(f"Counting: i is {i}")
+        i += 1
+
+demonstrate_break()
+```
+
+Possible example use cases:
+
+* needing to find a specific element, then leave the iteration
+* needing to leave the loop in case it comes across some invalid or error state
+
+`continue` is a keyword that will stop executing the current iteration, and resume execution at the next iteration. When our code sees `continue` , it doesn't run anymore code in that iteration, but continues to the next loop.
+
+``` python
+def demonstrate_continue():
+    i = 3
+
+    while i < 10:
+        i += 1
+        if i == 7:
+            print("i is seven!")
+            print("Let's print this special message, and then move on")
+            print("*~*~*~* \o/ *~*~*~*")
+            continue
+        print(f"Counting: i is {i}")
+
+demonstrate_continue()
+```
+
+Possible example uses cases:
+
+* wanting a loop to "skip" in a special case
+
+## Check for Understanding
+
+<!-- Need to figure out the best, most meaningful way to do a question in Learn -->
+
+<!-- question stubs -->
+
+Fix this loop so it uses the counter variable `i` , and runs while `i` is less than `5` .
+
+``` python
+def loop_until_five():
+    i = 0
+
+    while False:
+        i += 1
+
+    return i
+
+result = loop_until_five()
+```
+
+<!-- Answer: i should be 5 -->
+
+Which of these loop table options most accurately describes this loop?
+
+``` python
+i = 0
+result = None
+
+while i < 5:
+    if (i % 2) == 0:
+        result = True
+    else:
+        result = False
+    i += 1
+```
+
+<!-- Answer: This is the correct answer. False answers should toggle True/False on result, or be off-by-one on the i increments -->
+
+| Value of `i` at the beginning of the loop | `i < 5` ? | `i` at the end of loop body | `result` at end of loop body |
+|-------------------------------------------|----------|-----------------------------|------------------------------|
+| `0` | `True` | `1` | `True` |
+| `1` | `True` | `2` | `False` |
+| `2` | `True` | `3` | `True` |
+| `3` | `True` | `4` | `False` |
+| `4` | `True` | `5` | `True` |
+| `5` | `False` | Loop body doesn't execute   | -                            |
+
+Read this code. The function `look_for_treasure` should exit after we find treasure. Add in the `break` keyword in the right line to make this happen and pass the test.
+
+``` python
+import random
+
+def check_if_treasure(item):
+    is_treasure = random.randrange(100) < 25
+    # The line below uses the ternary operator.
+    return "Treasure!" if is_treasure else item + 1
+
+def look_for_treasure():
+    i = 0
+
+    while True:
+      if i == "Treasure!":
+        print("We found treasure instead of a number!")
+        print("Let's leave while we can.")
+      print(f"Counting: i is {i}")
+      i = check_if_treasure(i)
+
+result = look_for_treasure()
+```
+
+<!-- Answer: i should be "Treasure" -->
```

Comment:
```suggestion
### !challenge

* type: checkbox
* id: 891a2b93-62bd-4cd6-9b9f-d1be8d7da746
* title: Break keyword


##### !question

Read this code. The function `look_for_treasure` should exit after we find treasure. Select the option that has the `break` keyword in the right line to make this happen.

``` python
import random

def check_if_treasure(item):
    is_treasure = random.randrange(100) < 25
    # The line below uses the ternary operator.
    return "Treasure!" if is_treasure else item + 1

def look_for_treasure():
    i = 0

    while True:
      if i == "Treasure!":
        print("We found treasure instead of a number!")
        print("Let's leave while we can.")
      print(f"Counting: i is {i}")
      i = check_if_treasure(i)

result = look_for_treasure()
```

##### !end-question

##### !options

* 
```python
def check_if_treasure(item):
    is_treasure = random.randrange(100) < 25
    # The line below uses the ternary operator.
    return "Treasure!" if is_treasure else item + 1

def look_for_treasure():
    i = 0

    while True:
      if i == "Treasure!":
        print("We found treasure instead of a number!")
        print("Let's leave while we can.")
        break
      print(f"Counting: i is {i}")
      i = check_if_treasure(i)

result = look_for_treasure()
```

* 
```python
def check_if_treasure(item):
    is_treasure = random.randrange(100) < 25
    # The line below uses the ternary operator.
    return "Treasure!" if is_treasure else item + 1

def look_for_treasure():
    i = 0

    while True:
      if i == "Treasure!":
        print("We found treasure instead of a number!")
        print("Let's leave while we can.")
      print(f"Counting: i is {i}")
      break
      i = check_if_treasure(i)

result = look_for_treasure()
```

*
```python
def check_if_treasure(item):
    is_treasure = random.randrange(100) < 25
    # The line below uses the ternary operator.
    return "Treasure!" if is_treasure else item + 1

def look_for_treasure():
    i = 0

    while True:
      if i == "Treasure!":
        print("We found treasure instead of a number!")
        print("Let's leave while we can.")
      print(f"Counting: i is {i}")
      
      i = check_if_treasure(i)
      break

result = look_for_treasure()
```

##### !end-options

##### !answer

* 
```python
def check_if_treasure(item):
    is_treasure = random.randrange(100) < 25
    # The line below uses the ternary operator.
    return "Treasure!" if is_treasure else item + 1

def look_for_treasure():
    i = 0

    while True:
      if i == "Treasure!":
        print("We found treasure instead of a number!")
        print("Let's leave while we can.")
        break
      print(f"Counting: i is {i}")
      i = check_if_treasure(i)

result = look_for_treasure()
```


##### !end-answer

### !end-challenge

```

Reply:
I also might think this question would be better as part of an activity?